File: python/kohaku/src/kohaku/spec/diff.py
"""Compute UI Spec diffs and apply patches (port of TS diff.ts)."""

from __future__ import annotations

from typing import Any

from .canonical_json import canonical_stringify
from .errors import SpecError
from .models import ComponentNode, SpecPatch, UISpec
from .validate import ROOT_COMPONENT_ID, has_errors, validate_spec_structure


def order_components(components: list[ComponentNode]) -> list[ComponentNode]:
    """Order components in root-first DFS order (unreachable components sorted by id at the end).

    The canonical order of a Spec. Both composer's deterministic post-processing and apply_patch use it.
    """
    by_id = {c.id: c for c in components}
    ordered: list[ComponentNode] = []
    visited: set[str] = set()

    def visit(node_id: str) -> None:
        if node_id in visited:
            return
        node = by_id.get(node_id)
        if node is None:
            return
        visited.add(node_id)
        ordered.append(node)
        for child in node.children or []:
            visit(child)

    visit(ROOT_COMPONENT_ID)

    orphans = sorted((c for c in components if c.id not in visited), key=lambda c: c.id)
    return [*ordered, *orphans]


def _same_json(a: object, b: object) -> bool:
    """Drop models / None (equivalent to undefined) to wire form and take canonical equality."""
    return canonical_stringify(_to_plain(a)) == canonical_stringify(_to_plain(b))


def _to_plain(value: object) -> object:
    if value is None:
        return None
    to_wire = getattr(value, "to_wire", None)
    if callable(to_wire):
        return to_wire()
    if isinstance(value, list):
        return [_to_plain(v) for v in value]
    return value


def diff_spec(prev: UISpec, next_spec: UISpec) -> SpecPatch:
    """Compute the prev → next diff as a SpecPatch."""
    kwargs: dict[str, Any] = {"baseIntentHash": prev.intent.hash}

    if prev.kohaku != next_spec.kohaku:
        kwargs["kohaku"] = next_spec.kohaku
    if not _same_json(prev.intent, next_spec.intent):
        kwargs["intent"] = next_spec.intent

    prev_by_id = {c.id: c for c in prev.components}
    next_ids = {c.id for c in next_spec.components}

    upsert = [
        c
        for c in next_spec.components
        if c.id not in prev_by_id or not _same_json(prev_by_id[c.id], c)
    ]
    if upsert:
        kwargs["upsert"] = upsert

    remove = [c.id for c in prev.components if c.id not in next_ids]
    if remove:
        kwargs["remove"] = remove

    if not _same_json(prev.events, next_spec.events):
        kwargs["events"] = next_spec.events
    if prev.dataVersion != next_spec.dataVersion:
        kwargs["dataVersion"] = next_spec.dataVersion
    # Full replacement. If next no longer has refVersions (present only in prev), instruct removal with null.
    if not _same_json(prev.refVersions, next_spec.refVersions):
        kwargs["refVersions"] = next_spec.refVersions
    # state is a full replacement of the same kind (null = removal).
    if not _same_json(prev.state, next_spec.state):
        kwargs["state"] = next_spec.state
    if not _same_json(prev.provenance, next_spec.provenance):
        kwargs["provenance"] = next_spec.provenance

    return SpecPatch(**kwargs)


def apply_patch(prev: UISpec, patch: SpecPatch) -> UISpec:
    """Apply a patch. The resulting components are put in canonical order (order_components) and validated structurally.

    If prev is in canonical order, apply_patch(prev, diff_spec(prev, next)) equals next (in canonical order).
    """
    if patch.baseIntentHash != prev.intent.hash:
        raise SpecError(
            "PATCH_BASE_MISMATCH",
            f"patch targets intent {patch.baseIntentHash} but spec has {prev.intent.hash}",
        )

    by_id = {c.id: c for c in prev.components}
    for node_id in patch.remove or []:
        by_id.pop(node_id, None)
    for c in patch.upsert or []:
        by_id[c.id] = c

    kwargs: dict[str, Any] = {
        "kohaku": patch.kohaku if patch.kohaku is not None else prev.kohaku,
        "intent": patch.intent if patch.intent is not None else prev.intent,
        "dataVersion": patch.dataVersion if patch.dataVersion is not None else prev.dataVersion,
        "components": order_components(list(by_id.values())),
        "events": patch.events if patch.events is not None else prev.events,
        "provenance": patch.provenance if patch.provenance is not None else prev.provenance,
    }
    # Tri-state: key absent = no change (keep prev) / null = removal / dict = full replacement
    if patch.is_field_set("refVersions"):
        if patch.refVersions is not None:
            kwargs["refVersions"] = patch.refVersions
    elif prev.refVersions is not None:
        kwargs["refVersions"] = prev.refVersions
    if patch.is_field_set("state"):
        if patch.state is not None:
            kwargs["state"] = patch.state
    elif prev.state is not None:
        kwargs["state"] = prev.state

    next_spec = UISpec(**kwargs)

    issues = validate_spec_structure(next_spec)
    if has_errors(issues):
        raise SpecError(
            "PATCH_APPLY_FAILED",
            "patched spec is structurally invalid: "
            + ", ".join(i.code for i in issues if i.severity == "error"),
            issues,
        )
    return next_spec

File: packages/spec-core/src/diff.ts
import { canonicalStringify } from "./canonical-json.js";
import { SpecError } from "./errors.js";
import type { ComponentNode } from "./schema/component.js";
import type { JsonValue } from "./schema/json.js";
import type { SpecPatch } from "./schema/patch.js";
import type { UISpec } from "./schema/spec.js";
import { hasErrors, ROOT_COMPONENT_ID, validateSpecStructure } from "./validate.js";

// The source of truth for SpecPatch is the Zod schema (schema/patch.ts). To avoid breaking the
// existing import site (@kohaku-ui/spec-core), the type is re-exported from here. The diffSpec /
// applyPatch logic is unchanged.
export type { SpecPatch } from "./schema/patch.js";

/**
 * Orders components in root-first DFS order (unreachable components sorted by id at the end).
 * This is the Spec's canonical order. Both composer's deterministic post-processing and applyPatch
 * use it.
 */
export function orderComponents(components: ComponentNode[]): ComponentNode[] {
  const byId = new Map(components.map((c) => [c.id, c]));
  const ordered: ComponentNode[] = [];
  const visited = new Set<string>();

  const visit = (id: string): void => {
    if (visited.has(id)) return;
    const node = byId.get(id);
    if (!node) return;
    visited.add(id);
    ordered.push(node);
    for (const child of node.children ?? []) visit(child);
  };
  visit(ROOT_COMPONENT_ID);

  // The full comparator form that returns 0 on equality (aligned with the canonical-json.ts style).
  // IDs are unique, so behavior is unchanged.
  const orphans = components
    .filter((c) => !visited.has(c.id))
    .sort((a, b) => (a.id < b.id ? -1 : a.id > b.id ? 1 : 0));
  return [...ordered, ...orphans];
}

function sameJson(a: unknown, b: unknown): boolean {
  // Same reference (many components carry over the previous tree's reference unchanged when there is
  // no diff) returns true immediately without going through canonical serialization.
  if (a === b) return true;
  return canonicalStringify(a) === canonicalStringify(b);
}

/** How applyPatch applies refVersions (undefined = no change / null = remove / object = full replace). */
function nextRefVersions(prev: UISpec, patch: SpecPatch): { refVersions?: Record<string, string> } {
  if (patch.refVersions === undefined) {
    return prev.refVersions != null ? { refVersions: prev.refVersions } : {};
  }
  return patch.refVersions === null ? {} : { refVersions: patch.refVersions };
}

/** How applyPatch applies state (same shape as refVersions: undefined = no change / null = remove / object = full replace). */
function nextState(prev: UISpec, patch: SpecPatch): { state?: Record<string, JsonValue> } {
  if (patch.state === undefined) {
    return prev.state != null ? { state: prev.state } : {};
  }
  return patch.state === null ? {} : { state: patch.state };
}

/** Computes the prev → next diff as a SpecPatch. */
export function diffSpec(prev: UISpec, next: UISpec): SpecPatch {
  const patch: SpecPatch = { baseIntentHash: prev.intent.hash };

  if (prev.kohaku !== next.kohaku) patch.kohaku = next.kohaku;
  if (!sameJson(prev.intent, next.intent)) patch.intent = next.intent;

  const prevById = new Map(prev.components.map((c) => [c.id, c]));
  const nextIds = new Set(next.components.map((c) => c.id));

  const upsert = next.components.filter((c) => {
    const old = prevById.get(c.id);
    return old == null || !sameJson(old, c);
  });
  if (upsert.length > 0) patch.upsert = upsert;

  const remove = prev.components.filter((c) => !nextIds.has(c.id)).map((c) => c.id);
  if (remove.length > 0) patch.remove = remove;

  if (!sameJson(prev.events, next.events)) patch.events = next.events;
  if (prev.dataVersion !== next.dataVersion) patch.dataVersion = next.dataVersion;
  // Full replace. When next no longer has refVersions (only prev did), null signals removal.
  if (!sameJson(prev.refVersions, next.refVersions)) patch.refVersions = next.refVersions ?? null;
  // state is likewise a full replace (null = remove).
  if (!sameJson(prev.state, next.state)) patch.state = next.state ?? null;
  if (!sameJson(prev.provenance, next.provenance)) patch.provenance = next.provenance;

  return patch;
}

/**
 * Applies a patch. The resulting components are aligned to the canonical order (orderComponents) and
 * structurally validated. If prev is in canonical order, applyPatch(prev, diffSpec(prev, next))
 * equals next (in canonical order).
 */
export function applyPatch(prev: UISpec, patch: SpecPatch): UISpec {
  if (patch.baseIntentHash !== prev.intent.hash) {
    throw new SpecError(
      "PATCH_BASE_MISMATCH",
      `patch targets intent ${patch.baseIntentHash} but spec has ${prev.intent.hash}`,
    );
  }

  const byId = new Map(prev.components.map((c) => [c.id, c]));
  for (const id of patch.remove ?? []) byId.delete(id);
  for (const c of patch.upsert ?? []) byId.set(c.id, c);

  const next: UISpec = {
    kohaku: patch.kohaku ?? prev.kohaku,
    intent: patch.intent ?? prev.intent,
    dataVersion: patch.dataVersion ?? prev.dataVersion,
    // undefined = no change (keep prev) / null = remove / object = full replace
    ...nextRefVersions(prev, patch),
    ...nextState(prev, patch),
    components: orderComponents([...byId.values()]),
    events: patch.events ?? prev.events,
    provenance: patch.provenance ?? prev.provenance,
  };

  const issues = validateSpecStructure(next);
  if (hasErrors(issues)) {
    throw new SpecError(
      "PATCH_APPLY_FAILED",
      `patched spec is structurally invalid: ${issues
        .filter((i) => i.severity === "error")
        .map((i) => i.code)
        .join(", ")}`,
      issues,
    );
  }
  return next;
}
